src/validation/statement.rs

```diff
@@ -261,6 +262,45 @@ fn validate_direct_access<T: AnnotationMap>(
     }
 }
 
+fn validate_for_loop<T: AnnotationMap>(
+    validator: &mut Validator,
+    context: &ValidationContext<T>,
+    statement: &ForLoopStatement,
+) {
+    let mut reference: Option<(&DataType, AstNode)> = None;
+
+    let mut validate_if_datatype_is_numerical = |node: Option<&AstNode>| {
+        let Some(node) = node else { return };
+        let kind = context.annotations.get_type_or_void(node, context.index);
+
+        if kind.is_real() || !kind.is_numerical() {
+            let slice = get_datatype_name_or_slice(validator.context, kind);
+            let message = format!("Expected an integer value, got `{slice}`");
+            validator.push_diagnostic(
+                Diagnostic::error(message).with_location(node.get_location()).with_error_code("E093"),
+            );
+
+            return;
+        }
+
+        if reference.is_none() {
+            reference.replace((kind, node.clone()));
+        }
+    };
+
+    validate_if_datatype_is_numerical(Some(&statement.counter));
+    validate_if_datatype_is_numerical(Some(&statement.start));
+    validate_if_datatype_is_numerical(Some(&statement.end));
+    validate_if_datatype_is_numerical(statement.by_step.as_deref());
```

Comment:
You might be able to simplify this code by getting the conditionals from the `ForLoopStatement` (i.e. implement a `get_conditionals(&self) -> Vec<&AstNode>` function) and then iterate over the nodes, moving the logic into the iterator. E.g.:

```rust
statement.get_conditionals().iter().for_each(|node| {
        let kind = context.annotations.get_type_or_void(node, context.index);

        if kind.is_real() || !kind.is_numerical() {
            let slice = get_datatype_name_or_slice(validator.context, kind);
            let message = format!("Expected an integer value, got `{slice}`");
            validator.push_diagnostic(
                Diagnostic::error(message).with_location(node.get_location()).with_error_code("E093"),
            );
        }
    });
```